web/yo/app/scripts/controllers/gene.js

```diff
@@ -2514,17 +2514,78 @@ angular.module('oncokbApp')
+                var index = parent.indexOf(item);
+                switch (type) {
+                case 'top':
+                    if (index === 0) {
+                        return false;
+                    } else {
+                        return true;
+                    }
+                case 'bottom':
+                    if (index === parent.length - 1) {
+                        return false;
+                    } else {
+                        return true;
+                    }
+                case 'up':
+                    if (index === 0) {
+                        return false;
+                    } else {
+                        return true;
+                    }
+                case 'down':
+                    if (index === parent.length - 1) {
+                        return false;
+                    } else {
+                        return true;
+                    }
+                }
+            }
+            $scope.move = function (angleType, event, type, mutation, tumor, ti, treatment) {
                 var tmpStatus;
                 var moveStatusIndex;
                 var indexes = [];
-                var geneStatus = angular.copy($scope.geneStatus);
+                var geneStatus;
                 var key;
                 var numKey;
+                var driveList;
+                var index;
+                var moveIndex;
                 $scope.stopCollopse(event);
 
-                index = parseInt(index, 10);
-                moveIndex = parseInt(moveIndex, 10);
+                if (type === 'mutation') {
+                    geneStatus = $scope.geneStatus;
+                    driveList = this.gene.mutations;
+                    index = driveList.indexOf(mutation);
+                } else if (type === 'tumor') {
+                    var mutationIndex = this.gene.mutations.indexOf(mutation);
+                    geneStatus = $scope.geneStatus[mutationIndex];
+                    driveList = mutation.tumors;
+                    index = driveList.indexOf(tumor);
+                } else if (type === 'treatment') {
+                    var mutationIndex = this.gene.mutations.indexOf(mutation);
+                    var tumorIndex = mutation.tumors.indexOf(tumor);
+                    var tiIndex = tumor.TI.indexOf(ti);
+                    geneStatus = $scope.geneStatus[mutationIndex][tumorIndex][tiIndex];
+                    driveList = ti.treatments;
+                    index = driveList.indexOf(treatment);
+                }
+                switch(angleType) {
+                case 'up':
+                  moveIndex = index;
+                  break;
+                case 'down':
+                  moveIndex = index+2;
+                  break;
+                case 'top':
+                  moveIndex = 0;
+                  break;
+                case 'bottom':
+                  moveIndex = driveList.length;
+                  break;
+                }
```

Comment:
Technically, we should have a default as well. But optional from my opinion in this function.